File: packages/android-native/app/src/main/java/dev/tatliving/palmvellum/organizers/ui/screens/MailScreen.kt
package dev.tatliving.palmvellum.organizers.ui.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.navigation.NavHostController
import dev.tatliving.palmvellum.organizers.data.Graph
import dev.tatliving.palmvellum.organizers.data.Ulid
import dev.tatliving.palmvellum.organizers.data.local.RecordEntity
import dev.tatliving.palmvellum.organizers.data.model.MailFields
import dev.tatliving.palmvellum.organizers.data.model.MailSource
import dev.tatliving.palmvellum.organizers.data.model.PalmJson
import dev.tatliving.palmvellum.organizers.data.model.mailFieldsFrom
import dev.tatliving.palmvellum.organizers.ui.MasterDetailScaffold
import dev.tatliving.palmvellum.organizers.ui.PalmScaffold
import dev.tatliving.palmvellum.organizers.ui.components.EditorScaffold
import dev.tatliving.palmvellum.organizers.ui.components.PalmCategoryStrip
import dev.tatliving.palmvellum.organizers.ui.components.PalmDivider
import dev.tatliving.palmvellum.organizers.ui.components.PalmEmptyState
import dev.tatliving.palmvellum.organizers.ui.components.PalmField
import dev.tatliving.palmvellum.organizers.ui.components.PalmListCard
import dev.tatliving.palmvellum.organizers.ui.components.PalmRow
import dev.tatliving.palmvellum.organizers.ui.components.TitleAction
import dev.tatliving.palmvellum.organizers.ui.nav.Routes
import dev.tatliving.palmvellum.organizers.ui.theme.PalmInk
import dev.tatliving.palmvellum.organizers.ui.theme.PalmInkMute
import dev.tatliving.palmvellum.organizers.ui.theme.PalmRed
import dev.tatliving.palmvellum.organizers.ui.theme.PalmTitleBar
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import java.time.ZoneId

class MailViewModel : ViewModel() {
    private val repo = Graph.repo

    /** Inbox: AI "morning paper" digests delivered as type='mail' records. */
    val mails = repo.observeRecords("mail")
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val signedIn: Boolean get() = Graph.sync.isSignedIn

    fun refresh() = viewModelScope.launch { if (Graph.sync.isSignedIn) Graph.sync.syncNow() }

    /** Delete a mail digest from the inbox (soft-deletes the record + syncs). */
    fun delete(id: String) = viewModelScope.launch { repo.deleteRecord(id) }

    // ── mail_sources (subscriptions) live in their own table, online-only ──
    suspend fun loadSources(): Result<List<MailSource>> {
        val uid = Graph.session.userId ?: return Result.success(emptyList())
        return Graph.rest.select("mail_sources", "user_id=eq.$uid&order=created_at.desc").map { arr ->
            arr.mapNotNull { runCatching { PalmJson.decodeFromString<MailSource>(it.toString()) }.getOrNull() }
        }
    }

    suspend fun saveSource(src: MailSource): Result<Unit> =
        Graph.rest.upsert("mail_sources", listOf(PalmJson.encodeToJsonElement(src).jsonObject)).map { }

    suspend fun deleteSource(id: String): Result<Unit> =
        Graph.rest.delete("mail_sources", "id=eq.$id")
}

@Composable
fun MailScreen(navController: NavHostController) {
    val vm: MailViewModel = viewModel()
    val mails by vm.mails.collectAsState()
    LaunchedEffect(Unit) { vm.refresh() }

    var tab by remember { mutableStateOf("inbox") } // inbox | sources
    var openId by remember { mutableStateOf<String?>(null) }

    // Sources management is its own full screen, reachable via the title action.
    if (tab == "sources") {
        PalmScaffold(
            title = "Mail",
            navController = navController,
            currentRoute = Routes.MAIL,
            titleAction = { TitleAction("inbox") { tab = "inbox" } },
        ) { padding ->
            Column(Modifier.fillMaxSize().padding(padding)) { MailSources(vm) }
        }
        return
    }

    // Inbox: two-pane on Cosmo (open message left / inbox list right), classic
    // full-screen swap on standard — like the other list+detail screens.
    val open = openId?.let { id -> mails.firstOrNull { it.id == id } }
    MasterDetailScaffold(
        title = "Mail",
        navController = navController,
        currentRoute = Routes.MAIL,
        detail = open,
        titleAction = { TitleAction("sources") { tab = "sources" } },
        placeholder = "Pick a message from the inbox, or add a source.",
        master = { MailInbox(mails, signedIn = vm.signedIn, onOpen = { openId = it }) },
        detailContent = { rec, embedded ->
            MailRead(
                rec,
                embedded = embedded,
                onBack = { openId = null },
                onDelete = { vm.delete(rec.id); openId = null },
            )
        },
    )
}

@Composable
private fun MailInbox(mails: List<RecordEntity>, signedIn: Boolean, onOpen: (String) -> Unit) {
    Column(Modifier.fillMaxSize()) {
        if (!signedIn) {
            Text(
                "Sign in (Settings) to subscribe to sources and receive your AI morning paper.",
                color = PalmInkMute, fontSize = 13.sp,
                modifier = Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 8.dp),
            )
        }
        if (mails.isEmpty()) {
            PalmEmptyState("No mail yet. Add a subscription under 'sources' — your AI digest arrives at the time you set.")
        } else {
            val sorted = remember(mails) { mails.sortedByDescending { it.createdAt } }
            LazyColumn(modifier = Modifier.fillMaxSize().padding(10.dp)) {
                item {
                    PalmListCard {
                        sorted.forEachIndexed { i, rec ->
                            if (i > 0) PalmDivider()
                            val f = mailFieldsFrom(rec.metadataJson)
                            val unread = rec.aiStatus in listOf("pending", "processing", "queued")
                            PalmRow(
                                title = f.mail_subject ?: "(no subject)",
                                meta = f.mail_date_local ?: rec.createdAt.take(10),
                                body = listOfNotNull(f.mail_from ?: f.mail_source_name).joinToString().ifBlank { null },
                                dim = false,
                                metaColor = if (unread) PalmRed else PalmInkMute,
                                onClick = { onOpen(rec.id) },
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun MailRead(rec: RecordEntity, embedded: Boolean = false, onBack: () -> Unit, onDelete: () -> Unit) {
    val f = mailFieldsFrom(rec.metadataJson)
    EditorScaffold(title = "Mail", onCancel = onBack, saveEnabled = false, embedded = embedded, onSave = {}) {
        Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(14.dp)) {
            Text(f.mail_subject ?: "(no subject)", color = PalmInk, fontSize = 18.sp, fontWeight = FontWeight.Bold)
            Spacer(Modifier.height(4.dp))
            val from = listOfNotNull(f.mail_from ?: f.mail_source_name, f.mail_date_local).joinToString(" · ")
            if (from.isNotBlank()) Text(from, color = PalmInkMute, fontSize = 12.sp)
            if (f.mail_source_type == "topic" && !f.mail_topic.isNullOrBlank()) {
                Spacer(Modifier.height(4.dp))
                Text("Researched: ${f.mail_topic}", color = PalmInkMute, fontSize = 12.sp)
            }
            Spacer(Modifier.height(12.dp))
            Text(rec.body.orEmpty(), color = PalmInk, fontSize = 15.sp)
            val refs = f.mail_references.orEmpty()
            if (refs.isNotEmpty()) {
                Spacer(Modifier.height(16.dp))
                Text("REFERENCES", color = PalmTitleBar, fontSize = 11.sp, fontWeight = FontWeight.Bold)
                Spacer(Modifier.height(4.dp))
                refs.forEach { ref ->
                    Text("· $ref", color = PalmInkMute, fontSize = 12.sp, modifier = Modifier.padding(vertical = 1.dp))
                }
            }
            Spacer(Modifier.height(20.dp))
            DeleteButton(onDelete)
            Spacer(Modifier.height(24.dp))
        }
    }
}

@Composable
private fun MailSources(vm: MailViewModel) {
    val scope = rememberCoroutineScope()
    var sources by remember { mutableStateOf<List<MailSource>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }

    var name by remember { mutableStateOf("") }
    var type by remember { mutableStateOf("url") } // url | topic
    var url by remember { mutableStateOf("") }
    var topic by remember { mutableStateOf("") }
    var time by remember { mutableStateOf("07:00") }
    val tz = remember { runCatching { ZoneId.systemDefault().id }.getOrDefault("UTC") }
    var busy by remember { mutableStateOf(false) }

    fun reload() {
        scope.launch {
            loading = true
            vm.loadSources().fold(
                { sources = it; error = null },
                { error = it.message },
            )
            loading = false
        }
    }
    LaunchedEffect(Unit) { reload() }

    Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(vertical = 8.dp)) {
        Text(
            "Subscriptions",
            color = PalmInk, fontSize = 16.sp,
            modifier = Modifier.padding(horizontal = 14.dp, vertical = 4.dp),
        )
        Text(
            "A URL is fetched and summarised; a topic is researched. Your AI digest is delivered as mail at the fetch time, in your timezone.",
            color = PalmInkMute, fontSize = 13.sp,
            modifier = Modifier.padding(horizontal = 14.dp),
        )
        Spacer(Modifier.height(8.dp))

        when {
            loading -> Text("Loading...", color = PalmInkMute, fontSize = 13.sp, modifier = Modifier.padding(horizontal = 14.dp))
            error != null -> Text("Error: $error", color = PalmRed, fontSize = 13.sp, modifier = Modifier.padding(horizontal = 14.dp))
            sources.isEmpty() -> Text("No subscriptions yet.", color = PalmInkMute, fontSize = 13.sp, modifier = Modifier.padding(horizontal = 14.dp))
            else -> sources.forEach { src ->
                Row(
                    modifier = Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text(src.name, color = PalmInk, fontSize = 14.sp, maxLines = 1, overflow = TextOverflow.Ellipsis)
                        Text(
                            (if (src.source_type == "topic") src.topic else src.url).orEmpty() + "  ·  ${src.fetch_time.take(5)}",
                            color = PalmInkMute, fontSize = 11.sp, maxLines = 1, overflow = TextOverflow.Ellipsis,
                        )
                    }
                    Switch(
                        checked = src.enabled,
                        onCheckedChange = { on ->
                            scope.launch { vm.saveSource(src.copy(enabled = on)); reload() }
                        },
                    )
                    Text(
                        "remove",
                        color = PalmRed, fontSize = 13.sp,
                        modifier = Modifier.clickable {
                            scope.launch { vm.deleteSource(src.id); reload() }
                        }.padding(6.dp),
                    )
                }
            }
        }

        Spacer(Modifier.height(12.dp))
        PalmDivider()
        Spacer(Modifier.height(8.dp))
        Text("Add a subscription", color = PalmInk, fontSize = 15.sp, modifier = Modifier.padding(horizontal = 14.dp, vertical = 4.dp))
        PalmField("Name", name, { name = it })
        Text("Type", color = PalmInkMute, fontSize = 12.sp, modifier = Modifier.padding(start = 12.dp, top = 4.dp))
        PalmCategoryStrip(listOf("url" to "url", "topic" to "topic"), type) { type = it }
        if (type == "url") {
            PalmField("URL", url, { url = it })
        } else {
            PalmField("Topic", topic, { topic = it })
        }
        PalmField("Fetch time (HH:mm)", time, { time = it })
        Text("Timezone: $tz", color = PalmInkMute, fontSize = 12.sp, modifier = Modifier.padding(horizontal = 14.dp, vertical = 4.dp))
        Row(Modifier.padding(horizontal = 14.dp, vertical = 6.dp)) {
            val valid = name.isNotBlank() && (if (type == "url") url.isNotBlank() else topic.isNotBlank())
            Button(
                enabled = valid && !busy && vm.signedIn,
                onClick = {
                    busy = true
                    val t = time.trim().let { if (it.count { c -> c == ':' } == 1) "$it:00" else it }.ifBlank { "07:00:00" }
                    val src = MailSource(
                        id = Ulid.new(),
                        user_id = Graph.session.userId,
                        name = name.trim(),
                        source_type = type,
                        url = if (type == "url") url.trim() else null,
                        topic = if (type == "topic") topic.trim() else null,
                        fetch_time = t,
                        timezone = tz,
                        enabled = true,
                    )
                    scope.launch {
                        vm.saveSource(src).fold(
                            { name = ""; url = ""; topic = ""; error = null },
                            { error = it.message },
                        )
                        busy = false
                        reload()
                    }
                },
                colors = ButtonDefaults.buttonColors(containerColor = PalmTitleBar),
            ) { Text(if (busy) "Adding..." else "Add") }
        }
        Spacer(Modifier.height(24.dp))
    }
}

File: packages/mac-native/Sources/PalmVellum/Screens/MailScreen.swift
import SwiftUI
import PalmKit

struct MailScreen: View {
    @StateObject private var vm: MailVM
    @EnvironmentObject private var i18n: I18n
    @State private var tab = "Inbox"
    @State private var selection: String?

    init(repo: Repository, cloud: CloudClient) {
        _vm = StateObject(wrappedValue: MailVM(repo: repo, cloud: cloud))
    }

    var body: some View {
        PalmScaffold(
            title: i18n.t("mail"),
            titleAction: AnyView(deleteAllAction)
        ) {
            VStack(spacing: 0) {
                HStack {
                    PalmFilterStrip(options: ["Inbox", "Sources"], selection: $tab,
                                    labelFor: { i18n.t($0.lowercased()) })
                    Spacer()
                }
                .padding(.horizontal, 10).padding(.vertical, 6).background(Palm.desk)
                Divider()
                if tab == "Inbox" { inbox } else { sources }
            }
        }
        .task { await vm.loadSources() }
    }

    @ViewBuilder private var deleteAllAction: some View {
        if tab == "Inbox" && !vm.inbox.isEmpty {
            PalmTitleButton(label: i18n.t("deleteAll")) { vm.deleteAll(); selection = nil }
        } else { EmptyView() }
    }

    // MARK: Inbox

    private var inbox: some View {
        HStack(spacing: 0) {
            ScrollView {
                LazyVStack(spacing: 0) {
                    ForEach(vm.inbox) { mail in
                        let meta = PalmJSON.dict(mail.metadata)
                        Button(action: { selection = mail.id }) {
                            VStack(alignment: .leading, spacing: 2) {
                                Text(meta["mail_subject"] as? String ?? "(no subject)")
                                    .font(.system(size: 13)).foregroundColor(Palm.ink).lineLimit(1)
                                Text(meta["mail_from"] as? String ?? meta["mail_source_name"] as? String ?? "")
                                    .font(.system(size: 10)).foregroundColor(Palm.accentInk).lineLimit(1)
                            }
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 10).padding(.vertical, 7)
                            .background(selection == mail.id ? Palm.field : Color.clear)
                        }
                        .buttonStyle(.plain)
                        Divider()
                    }
                }
            }
            .frame(width: 260).background(Palm.desk)
            Divider()
            if let id = selection, let mail = vm.inbox.first(where: { $0.id == id }) {
                mailReader(mail)
            } else if vm.inbox.isEmpty {
                EmptyHint(text: "(inbox empty — add a source to receive AI digests)")
            } else {
                EmptyHint(text: "(select a message)")
            }
        }
    }

    private func mailReader(_ mail: PalmRecord) -> some View {
        let meta = PalmJSON.dict(mail.metadata)
        return ScrollView {
            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text(meta["mail_subject"] as? String ?? "(no subject)")
                        .font(.system(size: 15, weight: .semibold)).foregroundColor(Palm.ink)
                    Spacer()
                    DeleteButton { vm.delete(mail.id); selection = nil }
                }
                if let from = meta["mail_from"] as? String ?? meta["mail_source_name"] as? String {
                    Text(from).font(.system(size: 11)).foregroundColor(Palm.accentInk)
                }
                Divider()
                Text(mail.body ?? "").font(.system(size: 13)).foregroundColor(Palm.ink).textSelection(.enabled)
                if let refs = meta["mail_references"] as? [String], !refs.isEmpty {
                    Divider()
                    Text("references").font(.system(size: 10, weight: .semibold)).foregroundColor(Palm.accentInk)
                    ForEach(refs, id: \.self) { r in
                        Text(r).font(.system(size: 11)).foregroundColor(Palm.link).textSelection(.enabled)
                    }
                }
            }
            .padding(14)
        }
    }

    // MARK: Sources

    @ViewBuilder private var sources: some View {
        if !vm.isSignedIn {
            EmptyHint(text: "(sign in to manage mail subscriptions)")
        } else {
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    if let err = vm.sourcesError {
                        Text(err).font(.system(size: 10)).foregroundColor(Palm.offline)
                    }
                    ForEach(vm.sources) { src in
                        sourceRow(src)
                    }
                    Divider()
                    AddSourceForm { push in Task { await vm.saveSource(push) } }
                }
                .padding(14).frame(maxWidth: 520).frame(maxWidth: .infinity)
            }
        }
    }

    private func sourceRow(_ src: MailSource) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack {
                Text(src.name).font(.system(size: 13, weight: .semibold)).foregroundColor(Palm.ink)
                Spacer()
                Text(src.source_type).font(.system(size: 10)).foregroundColor(Palm.accentInk)
            }
            if let u = src.url, !u.isEmpty { Text(u).font(.system(size: 10)).foregroundColor(Palm.link).lineLimit(1) }
            if let t = src.topic, !t.isEmpty { Text("topic: \(t)").font(.system(size: 10)).foregroundColor(Palm.accentInk) }
            HStack(spacing: 12) {
                Text(src.enabled ? "enabled" : "disabled").font(.system(size: 10)).foregroundColor(Palm.accentInk)
                Button("fetch now") { Task { await vm.fetchNow(src.id) } }.buttonStyle(.plain).foregroundColor(Palm.link).font(.system(size: 11))
                Button("delete") { Task { await vm.deleteSource(src.id) } }.buttonStyle(.plain).foregroundColor(Palm.offline).font(.system(size: 11))
            }
        }
        .palmCard()
    }
}

private struct AddSourceForm: View {
    let onAdd: (MailSourcePush) -> Void
    @State private var name = ""
    @State private var type = "url"
    @State private var url = ""
    @State private var topic = ""
    @State private var language = "auto"

    private let languages = ["auto", "English", "繁體中文", "简体中文", "日本語", "한국어", "Русский"]

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("add subscription").font(.system(size: 12, weight: .semibold)).foregroundColor(Palm.ink)
            PalmLabeledField(label: "name", text: $name)
            Picker("", selection: $type) { Text("url").tag("url"); Text("topic").tag("topic") }
                .pickerStyle(.segmented).labelsHidden()
            if type == "url" {
                PalmLabeledField(label: "feed url", text: $url, placeholder: "https://…")
            } else {
                PalmLabeledField(label: "topic", text: $topic, placeholder: "what to research")
            }
            HStack {
                Text("language").font(.system(size: 11, weight: .semibold)).foregroundColor(Palm.accentInk)
                Picker("", selection: $language) { ForEach(languages, id: \.self) { Text($0).tag($0) } }.labelsHidden()
            }
            Button("add") { add() }.buttonStyle(.plain).foregroundColor(Palm.link)
                .disabled(name.isEmpty || (type == "url" ? url.isEmpty : topic.isEmpty))
        }
        .palmCard()
    }

    private func add() {
        let push = MailSourcePush(
            id: Ulid.new(), user_id: "", name: name,
            url: type == "url" ? url : nil, topic: type == "topic" ? topic : nil,
            source_type: type, enabled: true, fetch_time: "07:00:00", timezone: "Asia/Hong_Kong",
            output_language: language == "auto" ? nil : language)
        onAdd(push)
        name = ""; url = ""; topic = ""
    }
}
